### PR TITLE
net: lib: coap: Change coap_pending_init()

### PR DESCRIPTION
**Pull request by juhaylinen**
Replace function parameter 'retries' with pointer to structure holding coap transmission parameters. This allows setting the retransmission parameters individually for each pending request.

Add coap transmission parameters to coap_pending structure.

Update migration guide and release notes.

### CODE REVIEW

File: doc/releases/release-notes-3.6.rst [hunk 1]
```diff
@@ -212,6 +212,10 @@ Networking
 * CoAP:
 
   * Emit observer/service network events using the Network Event subsystem.
+  * Added new API functions:
```
Comment (pdgendt) [inline doc/releases/release-notes-3.6.rst:215]: Nit: add a newline between list items.

File: tests/net/lib/coap/src/main.c
Comment (pdgendt) [inline tests/net/lib/coap/src/main.c:1]: Can you add actual testing of the parameters?

File: include/zephyr/net/coap.h [hunk 2]
```diff
@@ -989,14 +1002,15 @@ void coap_reply_init(struct coap_reply *reply,
  * confirmation message, initialized with data from @a request
  * @param request Message waiting for confirmation
  * @param addr Address to send the retransmission
- * @param retries Maximum number of retransmissions of the message.
+ * @param params Pointer to the CoAP transmission parameters struct,
+ * or NULL to use default values set by Kconfig
```
Comment (rlubos) [inline include/zephyr/net/coap.h:1006]: nit: I'd drop the `set by Kconfig` part, as it's not necessarily correct - the initial Kconfig values could be overwritten by `coap_set_transmission_parameters()`.

File: tests/net/lib/coap/src/main.c
Comment (juhaylinen) [reply tests/net/lib/coap/src/main.c:1]: Added new test case for the parameters.

File: tests/net/lib/coap/src/main.c [hunk 3]
```diff
@@ -1718,4 +1718,55 @@ ZTEST(coap, test_coap_packet_set_path)
 						  COAP_OPTION_URI_PATH);
 }
 
+ZTEST(coap, test_transmission_parameters)
```
Comment (pdgendt) [inline tests/net/lib/coap/src/main.c:1721]: Fix the compliance issues and add a separate commit for this test.

File: subsys/net/lib/coap/coap_server.c [hunk 4]
```diff
@@ -550,8 +552,9 @@ int coap_service_send(const struct coap_service *service, const struct coap_pack
 			goto send;
 		}
 
-		ret = coap_pending_init(pending, cpkt, addr,
-					CONFIG_COAP_SERVICE_PENDING_RETRANSMITS);
+		params = coap_get_transmission_parameters();
+		params.max_retransmission = CONFIG_COAP_SERVICE_PENDING_RETRANSMITS;
+		ret = coap_pending_init(pending, cpkt, addr, &params);
```
[the remark this answers is not in the document]
Comment (juhaylinen) [reply subsys/net/lib/coap/coap_server.c:557]: @pdgendt feel free to make the changes.

**Review by pdgendt — DISMISSED** [review]

**Review by rlubos — DISMISSED** [review]

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

**Review by rlubos — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Pull request merged**